librz/analysis/arch/sh/sh_il.c

```diff
@@ -0,0 +1,1386 @@
+	RzILOpPure *msb = MSB(sh_il_get_pure_param(0));
+	RzILOpPure *shl = SHIFTL0(sh_il_get_pure_param(0), SH_U_REG(1));
+	return SEQ2(SETG(SH_SR_T, msb), sh_il_set_pure_param(0, shl));
+}
+
+/**
+ * SHLR  Rn
+ * 0 -> Rn -> T
+ * 0100nnnn00000001
+ */
+static RzILOpEffect *sh_il_shlr(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	RzILOpPure *lsb = LSB(sh_il_get_pure_param(0));
+	RzILOpPure *shr = SHIFTR0(sh_il_get_pure_param(0), SH_U_REG(1));
+	return SEQ2(SETG(SH_SR_T, lsb), sh_il_set_pure_param(0, shr));
+}
+
+/**
+ * SHLL2  Rn
+ * Rn << 2 -> Rn
+ * 0100nnnn00001000
+ */
+static RzILOpEffect *sh_il_shll2(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return sh_il_set_pure_param(0, SHIFTL0(sh_il_get_pure_param(0), SH_U_REG(2)));
+}
+
+/**
+ * SHLR2  Rn
+ * Rn >> 2 -> Rn
+ * 0100nnnn00001001
+ */
+static RzILOpEffect *sh_il_shlr2(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return sh_il_set_pure_param(0, SHIFTR0(sh_il_get_pure_param(0), SH_U_REG(2)));
+}
+
+/**
+ * SHLL8  Rn
+ * Rn << 8 -> Rn
+ * 0100nnnn00011000
+ */
+static RzILOpEffect *sh_il_shll8(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return sh_il_set_pure_param(0, SHIFTL0(sh_il_get_pure_param(0), SH_U_REG(8)));
+}
+
+/**
+ * SHLR8  Rn
+ * Rn >> 8 -> Rn
+ * 0100nnnn00011001
+ */
+static RzILOpEffect *sh_il_shlr8(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return sh_il_set_pure_param(0, SHIFTR0(sh_il_get_pure_param(0), SH_U_REG(8)));
+}
+
+/**
+ * SHLL16  Rn
+ * Rn << 16 -> Rn
+ * 0100nnnn00101000
+ */
+static RzILOpEffect *sh_il_shll16(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return sh_il_set_pure_param(0, SHIFTL0(sh_il_get_pure_param(0), SH_U_REG(16)));
+}
+
+/**
+ * SHLR16  Rn
+ * Rn >> 16 -> Rn
+ * 0100nnnn00101001
+ */
+static RzILOpEffect *sh_il_shlr16(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return sh_il_set_pure_param(0, SHIFTR0(sh_il_get_pure_param(0), SH_U_REG(16)));
+}
+
+/**
+ * CLRMAC
+ * 0 -> MACH, MACL
+ * 0000000000101000
+ */
+static RzILOpEffect *sh_il_clrmac(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return SEQ2(SETG("mach", UN(SH_REG_SIZE, 0)), SETG("macl", UN(SH_REG_SIZE, 0)));
+}
+
+/**
+ * CLRS
+ * 0 -> S
+ * 0000000001001000
+ */
+static RzILOpEffect *sh_il_clrs(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return SETG(SH_SR_S, IL_FALSE);
+}
+
+/**
+ * CLRT
+ * 0 -> T
+ * 0000000000001000
+ */
+static RzILOpEffect *sh_il_clrt(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	return SETG(SH_SR_T, IL_FALSE);
+}
+
+// TODO: This needs to be fixed for banked register LDC
+/**
+ * LDC  Rm, REG
+ * REG := SR/GBR/VBR/SSR/SPC/DBR/Rn_BANK
+ * Rm -> REG
+ * PRIVILEGED (Only GBR is not privileged)
+ *
+ * LDC.L  @Rm+, REG
+ * REG := SR/GBR/VBR/SSR/SPC/DBR/Rn_BANK
+ * (Rm) -> REG ; Rm + 4 -> Rm
+ * PRIVILEGED (Only GBR is not privileged)
+ */
+static RzILOpEffect *sh_il_ldc(SHOp *op, ut64 pc, RzAnalysis *analysis) {
+	RzBitVector *priv_bit = rz_il_evaluate_bitv(analysis->il_vm->vm, VARG(SH_SR_D));
+	ut8 state = priv_bit->bits.small_u == 0 ? 0b1 : 0b0;
```

Comment:
Standard C does not support binary literals, which is so wrong, but yeah, better to avoid custom extensions.

https://stackoverflow.com/questions/18244726/why-doesnt-c-have-binary-literals